make-live.8
.TH MAKE-LIVE 8 "Jul, 27 2006" "0.99" "Debian Live framework"

.SH NAME
make-live \- An utility for building Debian Live systems.

.SH SYNOPSIS
.B make-live
.RB [\| \-a \||\| \-\-architecture
.IR ARCHITECTURE \|]
.RB [\| \-c \||\| \-\-conffile
.IR FILE \|]
.RB [\| \-\-clone
.IR DIRECTORY \|]
.RB [\| \-d \||\| \-\-distribution
.IR DISTRIBUTION \|]
.BR [\| \-\-debug \|]
.RB [\| \-f \||\| \-\-filesystem
.IR FILESYSTEM \|]
.RB [\| \-\-flavour
.IR FLAVOUR \|]
.RB [\| \-\-hook
.RI \|" COMMAND .\|.\|.\|"\|]
.RB [\| \-\-include\-image
.IR FILE \||\| DIRECTORY \|]
.RB [\| \-\-include\-rootfs
.IR FILE \||\| DIRECTORY \|]
.RB [\| \-\-linux\-image
.IR FLAVOUR \|]
.RB [\| \-\-logfile
.IR FILE \|]
.RB [\| \-p \||\| \-\-package-list
.IR FILE \|]
.RB [\| \-\-preseed
.IR FILE \|]
.RB [\| \-\-rootfs
.IR DIRECTORY \|]
.RB [\| \-s \||\| \-\-section
.RI \|" SECTION \|.\|.\|.\|"\|]
.RB [\| \-\-splashy
.RI [\| THEME \|]\|]
.RB [\| \-t \||\| \-\-type
.IR TYPE \|]
.RB [\| \-\-templates
.IR DIRECTORY \|]
.RB [\| \-\-verbose \|]
.RI [\| LIST \|]
.PP
.B make-live
.RB [\| \-h \||\| \-\-help \|]
.br
.B make-live
.RB [\| \-u \||\| \-\-usage \|]
.br
.B make-live
.RB [\| \-v \||\| \-\-version \|]

.SH DESCRIPTION
The Debian Live framework lets you automagically create live system images.
.PP

.SH OPTIONS
.TP
.BI "\-a, \-\-architecture " ARCHITECTURE
Build an image for another architecture than the autodetected one. This is
usefull for e.g. building i386 images on an amd64 machine. Instead of this flag,
the environment variable LIVE_ARCHITECTURE can be used.
.TP
.BI "\-c, \-\-conffile " FILE
Uses another configuration file in addition to the default one, if existing, in
/etc/make-live.conf. Settings in the additional configuration file do overwrite
the settings from /etc/make-live.conf. Instead of this flag, the environment
varibale LIVE_CONFFILE can be used.
.TP
.BI "\-\-clone " DIRECTORY
Clones the package selection and configuration from a given root directory. It
will output the settings to $LIVE_ROOT/preseed.cloned and $LIVE_ROOT/plist.clone
for easy reuse with \-\-preseed and \-\-package-list. Instead of this flag, the
environment variable LIVE_CLONE can be used.
.TP
.BI "\-d, \-\-distribution " DISTRIBUTION
Specifies the distribution to be used. Allowed values are all Debian
distributions (oldstable, stable, testing, unstable), although only unstable
(default) is possible for now. Instead of this flag, the environment variable
LIVE_DISTRIBUTION can be used.
.TP
.B \-\-debug
Enables debut output of make-live itself (this includes \-\-verbose too).
Instead of this environment variable, LIVE_DEBUG can be used.
.TP
.BI "\-f, \-\-filesystem " FILESYSTEM
Specifies the filesystem for the compressed image. At the moment, only squashfs
(default) and ext2 are allowed values. Instead of this flag, the environment
variable LIVE_FILESYSTEM can be used.
.TP
.BI "\-\-flavour " FLAVOUR
Specifies the bootstrap flavour. This can be either standard (default) or
minimal. Instead of this flag, the environment variable LIVE_FLAVOUR can be
used.
.TP
.\" FIXME
.BI "\-\-hook " COMMAND "\|.\|.\|.\|"
Specifies a command which will be executed at the end of the chroot setup and
before the mastering of the image. Instead of this flag, the environment
variable LIVE_HOOK can be used.
.TP
.BI "\-\-include\-image " FILE \||\| DIRECTORY
specifies a file or directory which will be copied inside the image. Can be also
set through LIVE_INCLUDE_IMAGE environment variable.
.TP
.BI "\-\-include\-rootfs " FILE \||\| DIRECTORY
specifies a file or directory which will be copied inside the root filesystem of
the image. Can be also set through the LIVE_INCLUDE_ROOTFS environment variable.
.TP
.BI "\-\-linux\-image " FLAVOUR
.TP
.BI "\-p, \-\-package-list " FILE
specifies a package list file, one package for each line. It overrides list option.
.TP
.BI "\-\-logfile " FILE
specifies the location of the logfile. It can be set through LIVE_LOGFILE,
default is \$LIVE_ROOT/make-live.log
.TP
.BI "\-\-preseed " FILE
clone the packages selection and configuration from the provided
root directory.
.TP
.BI "\-\-rootfs " DIRECTORY
skip all phases of rootfs generation and just install casper and the
kernel in the specified root directory (must be a debian).
.TP
.BI "\-s, \-\-section " SECTION "\|.\|.\|.\|"
specifies which sections are going to be availble for the package selection inside the chroot (Default: main). Can be set to every combination of "main contrib non-free", also through LIVE_SECTIONS.
.TP
.BI "\-\-splashy [\|" THEME \|]
specifies that splashy should be installed and included in initamfs, an optional
official THEME could be selected or an exernal theme could be installed, in the
second case the .tar.gz theme should be passed without extension and must be in
the current directory. Can be also set through LIVE_SPLASHY and
LIVE_SPLASHY_THEME environment variables.
.TP
.BI "\-t, \-\-type " TYPE
specifies the image type to generate. Currently, only iso and net are available.
.TP
.BI "\-\-templates " DIRECTORY
specifies a templates directory to use instead of the default one.
.TP
.B \-\-verbose
providing debug information for third-party tools.
.TP
.B \-h, \-\-help
display this help and exit.
.TP
.B \-u, \-\-usage
show usage and exit.
.TP
.B \-v, \-\-version
output version information and exit.

.SH LISTS
Every list does already include the Debian standard package selection.
.TP
.B gnome | gnome-core | gnome-full
include GNOME.
.TP
.B kde | kde-core | kde-full | kde-extra
include KDE.
.TP
.B x11 | x11-core
include X-Window-System only.
.TP
.B xfce
include Xfce.

.SH BUGS
.B make-live
works fully on amd64 and i386, partially on alpha, hppa, ia64, mips, mipsel,
powerpc, s390 and sparc. Other architectures, such as arm and m68k, as well as
non-linux ports of Debian, like hurd-i386, are currently not supported.

.SH HOMEPAGE
Debian Live project <http://live.debian.net/>

.SH SEE ALSO
.BR make-live.conf (8).

.SH AUTHOR
make-live was written by Daniel Baumann <daniel@debian.org> and Marco Amadori
<marco.amadori@gmail.com>.


HISTORY
Milestones of Debian Live
-------------------------

2006-02-14: Initial idea
http://blog.daniel-baumann.ch/2006/02/14#20060214_debian-live-initiative

2006-03-16: First CD images available
http://lists.alioth.debian.org/pipermail/debian-live-devel/2006-March/000048.html

2006-06-02: First netboot images available

2007-02-05: Image autobuilder in place
http://lists.alioth.debian.org/pipermail/debian-live-devel/2007-February/000764.html

2007-02-10: First USB images available
http://lists.alioth.debian.org/pipermail/debian-live-devel/2007-February/000790.html


make-live.sh
#!/bin/sh

# make-live - An utility for building Debian Live systems.
#
# Copyright (C) 2006 Daniel Baumann <daniel@debian.org>
# Copyright (C) 2006 Marco Amadori <marco.amadori@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

set -e

PROGRAM="`basename ${0}`"
VERSION="0.99"

# Source default configuration
if [ -r /etc/make-live.conf ]
then
	. /etc/make-live.conf
fi

# Source alternate configuration
if [ ! -z "${LIVE_CONFFILE}" ]
then
	if [ -r "${LIVE_CONFFILE}" ]
	then
		. "${LIVE_CONFFILE}"
	else
		echo "W: could not read ${LIVE_CONFFILE}, using defaults."
	fi
fi

# Source scriptlets
for SCRIPT in /usr/share/make-live/scripts/*.sh
do
	. "${SCRIPT}"
done

Help ()
{
	echo "make-live - An utility for building Debian Live systems."
	echo
	echo "Usage: ${PROGRAM} [-a ARCHITECTURE] [-c FILE] [--clone DIRECTORY] [-d DISTRIBUTION] [--debug] [-f FILESYSTEM] [--flavour FLAVOUR] [--hook \"COMMAND...\"] [--include-image FILE|DIRECTORY] [--include-rootfs FILE|DIRECTORY] [--interactive] [--linux-image FLAVOUR] [--logfile FILE] [-p|--package-list FILE] [--preseed FILE] [--rootfs DIRECTORY] [-s|--section \"SECTION...\"] [--splashy [THEME]] [-t|--type TYPE] [--templates DIRECTORY] [--verbose] [LIST]"
	echo
	echo "Values:"
	echo "  Architectures: alpha, amd64, arm, hppa, i386, ia64, m68k, powerpc, s390, sparc."
	echo "  Distributions: oldstable, stable, testing, unstable."
	echo "  Filesystems: ext2, squashfs, xfs."
	echo "  Flavours: bootable, build, minimal, standard."
	echo "  Linux Images: alpha-generic, alpha-smp, alpha-legacy, amd64-k8, amd64-k8-smp, em64t-p4, em64t-p4-smp, footbridge, ixp4xx, nslu2, rpc, s3c2410, parisc, parisc-smp, parisc64, parisc64-smp, 486, 686, k7, itanium, itanium-smp, mckinley, mckinley-smp, amiga, mac, r4k-ip22, r5k-ip32, sb1-bcm91250a, sb1a-bcm91480b, r5k-cobalt, r3k-kn02, r4k-kn04, powerpc, powerpc-smp, powerpc-miboot, powerpc64, s390, s390x, sparc32, sparc64, sparc64-smp."
	echo "  Sections: main, contrib, non-free."
	echo "  Types: iso, net."
	echo "  Lists: gnome, gnome-core, gnome-full, kde, kde-core, kde-core-i18n, kde-extra, kde-extra-i18n, kde-full, kde-full-i18n, kde-18n, standard-i18n, x11, x11-core, xfce."
	echo
	echo "Options:"
	echo "  -h, --help:    display this help and exit."
	echo "  -u, --usage:   show usage and exit."
	echo "  -v, --version: output version information and exit."
	echo
	echo "Environment:"
	echo "  All settings can also be done through environment variables. Please see make-live.conf(8) for more information."
	echo
	echo "Report bugs to Debian Live project <http://live.debian.net>."
	exit 1
}

Usage ()
{
	echo "make-live - An utility for building Debian Live systems."
	echo
	echo "Usage: ${PROGRAM} [-a ARCHITECTURE] [-c FILE] [--clone DIRECTORY] [-d DISTRIBUTION] [--debug] [-f FILESYSTEM] [--flavour FLAVOUR] [--hook "COMMAND..."] [--include-image FILE|DIRECTORY] [--include-rootfs FILE|DIRECTORY] [--linux-image FLAVOUR] [--logfile FILE] [-p|--package-list FILE] [--preseed FILE] [--rootfs DIRECTORY] [-s|--section "SECTION..."] [--splashy [THEME]] [-t|--type TYPE] [--templates DIRECTORY] [--verbose] [LIST]"
	echo
	echo "Try \"${PROGRAM} --help\" for more information."
	exit 1
}

Version ()
{
	echo "make-live, version ${VERSION}"
	echo
	echo "Copyright (C) 2006 Daniel Baumann <daniel@debian.org>"
	echo "Copyright (C) 2006 Marco Amadori <marco.amadori@gmail.com>"
	echo
	echo "This program is free software; you can redistribute it and/or modify"
	echo "it under the terms of the GNU General Public License as published by"
	echo "the Free Software Foundation; either version 2 of the License, or"
	echo "(at your option) any later version."
	echo
	echo "This program is distributed in the hope that it will be useful,"
	echo "but WITHOUT ANY WARRANTY; without even the implied warranty of"
	echo "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the"
	echo "GNU General Public License for more details."
	echo
	echo "Homepage: Debian Live project <http://live.debian.net/>"
	exit 1
}

Main ()
{
	ARGUMENTS=`getopt --shell=sh --name="${PROGRAM}" \
		--options="a:c:d:f:p:s:t:huv" \
		--longoptions="architecture:,conffile:,clone:,distribution:,debug,filesystem:,flavour:,hook:,include-image:,include-rootfs:,interactive,linux-image:,logfile:,package-list:,preseed:,rootfs:,section:,splashy::,type:,templates:,verbose,help,usage,version" \
		-- "${@}"`

	if [ "${?}" != "0" ]
	then
		echo "Terminating..." >&2
		exit 1
	fi

	eval set -- "${ARGUMENTS}"

	while true
	do
		case "${1}" in
			-a|--architecture)
				LIVE_ARCHITECTURE="${2}"; shift 2
				;;

			-c|--conffile)
				LIVE_CONFIG="${2}"; shift 2
				;;

			--clone)
				LIVE_CLONE="${2}"; shift 2
				;;

			-d|--distribution)
				LIVE_DISTRIBUTION="${2}"; shift 2
				;;

			--debug)
				LIVE_DEBUG="1"
				LIVE_VERBOSE="1"; shift
				;;

			-f|--filesystem)
				LIVE_FILESYSTEM="${2}"; shift 2
				;;

			--flavour)
				LIVE_FLAVOUR="${2}"; shift 2
				;;

			--hook)
				LIVE_HOOK="${2}"; shift 2
				;;

			--include-image)
				LIVE_INCLUDE_IMAGE="${2}"; shift 2
				;;

			--include-rootfs)
				LIVE_INCLUDE_ROOTFS="${2}"; shift 2
				;;

			--interactive)
				LIVE_INTERACTIVE="1"; shift
				;;

			--linux-image)
				LIVE_LINUX="${2}"; shift 2
				;;

			--logfile)
				LIVE_LOGFILE="${2}"; shift 2
				;;

			-p|--package-list)
				LIVE_PACKAGE_LIST="${2}"; shift 2
			    ;;

			--preseed)
				LIVE_PRESEED="${2}"; shift 2
				;;

			--rootfs)
				LIVE_ROOTFS="${2}"; shift 2
				;;

			-s|--section)
				LIVE_SECTIONS="${2}"; shift 2
				;;

			--splashy)
				LIVE_SPLASHY="1" 
				case "${2}" in
					"") 
						shift 2;
						;;
					*)
						LIVE_SPLASHY_THEME=${2} ; shift 2
						;;
				esac
				;;

			-t|--type)
				LIVE_TYPE="${2}"; shift 2
				;;

			--templates)
				LIVE_TEMPLATES="${2}"; shift 2
				;;

			--verbose)
				LIVE_VERBOSE="1"; shift
				;;

			-h|--help)
				Help; shift
				;;

			-u|--usage)
				Usage; shift
				;;

			-v|--version)
				Version; shift
				;;

			--)
				shift; break
				;;

			*)
				echo "Internal error!"
				exit 1
				;;
		esac
	done

	LIVE_LIST="${1}"

	# Enabling debug
	if [ "${LIVE_DEBUG}" ]
	then
		set +x
	fi

	# Setting defaults
	Defaults

	# Initial checks
	Init

	# Bootstrap chroot
	Bootstrap

	# Customize chroot
	Chroot

	# Create type
	if [ "${LIVE_TYPE}" ]
	then
		case ${LIVE_TYPE} in
			iso)
				LIVE_TYPE="Iso"
				;;

			net)
				LIVE_TYPE="Net"
				;;

			*)
				echo "You specified a wrong image type"
				Help
				;;
		esac

		${LIVE_TYPE}
	else
		Iso
	fi
}

Main "${@}"
